### PR TITLE
Support layers without parameters or state

### PR DESCRIPTION
**Pull request by gdalle**
This PR proposes a solution to #83 by adding 2 booleans `P` and `S` to the layer types, specifying whether or not they need parameters / states as inputs.

That way, in addition to the generic syntax `layer(x, ps, st)`:
- a layer without parameters (e.g. `Dropout`) supports `layer(x, st)`
- a layer without state (e.g. `Dense`) supports `layer(x, ps)`
- a layer without parameters or state (e.g. `ReshapeLayer`) supports `layer(x)`

### CODE REVIEW

**Comment by gdalle** [comment]
Sofar I haven't added specific tests for these new methods, and they are not documented either. This is just a first draft to agree on the syntax before moving forward

**Comment by gdalle** [comment]
The main issue I see is for constructs such as `Parallel`, `BranchLayer`, `PairwiseFusion` and `Chain`. Since those layers are built from potentially heterogeneous `NamedTuple`s, I don't know how to statically determine whether they have parameters / states. One would need to apply an OR to the `hasparams` / `hasstate` of each element of the `NamedTuple`, which sounds like it has to be done dynamically since we don't even know its length.

I was thinking, would be recursive pairwise construction work? If a chain is stored as `Chain(Chain(l1, l2), l3)` instead of `Chain(l1, l2, l3)`, does that make it easier to propagate parameter / state information? Does it slow computations down?

**Comment by gdalle** [comment]
> The main issue I see is for constructs such as `Parallel`, `BranchLayer`, `PairwiseFusion` and `Chain`.

I may have found a solution. A `Chain` (or any other similar construct) requires parameters / states if and only if at least one of its components does. If I'm right, the code below can decide that statically (just imagine `B` is either `hasparams` or `hasstate`):

```julia
julia> using Test

julia> struct MyLayer{B} end

julia> struct MyChain{B} end

julia> MyChain(t::Tuple{Vararg{MyLayer}}) = MyChainTuple{true}();

julia> MyChain(t::Tuple{Vararg{MyLayer{false}}}) = MyChainTuple{false}();

julia> l0, l1 = MyLayer{false}(), MyLayer{true}();

julia> c00 = @inferred MyChain((l0, l0))
MyChainTuple{false}()

julia> c01 = @inferred MyChain((l0, l1))
MyChainTuple{true}()

julia> c11 = @inferred MyChain((l1, l1))
MyChainTuple{true}()
```

File: src/core.jl [hunk 1]
```diff
@@ -20,14 +25,18 @@ Optionally:
...
+abstract type AbstractExplicitLayer{hasparams, hasstate} end
 
 """
     initialparameters(rng::AbstractRNG, l)
 
 Generate the initial parameters of the layer `l`.
 """
 initialparameters(::AbstractRNG, ::AbstractExplicitLayer) = NamedTuple()
+function initialparameters(::AbstractRNG,
+                           ::AbstractExplicitLayer{false, hasstate}) where {hasstate}
```
Comment (avik-pal) [inline src/core.jl:37]: Unused `hasstate`

File: src/core.jl [hunk 2]
```diff
@@ -48,6 +61,9 @@ Return the total number of parameters of the layer `l`.
 function parameterlength(l::AbstractExplicitLayer)
     return parameterlength(initialparameters(Random.default_rng(), l))
 end
+function parameterlength(::AbstractExplicitLayer{hasstate, false}) where {hasstate}
```
Comment (avik-pal) [inline src/core.jl:64]: Ordering is incorrect

File: src/core.jl [hunk 3]
```diff
@@ -20,14 +25,18 @@ Optionally:
 
 See also [`AbstractExplicitContainerLayer`](@ref)
 """
-abstract type AbstractExplicitLayer end
+abstract type AbstractExplicitLayer{hasparams, hasstate} end
```
Comment (avik-pal) [inline src/core.jl:28]: Let's add functions to check this as well. `hasstates(::AbstractExplicitLayer)` and `hasparams(::AbstractExplicitLayer)`

File: src/core.jl [hunk 4]
```diff
@@ -59,6 +75,7 @@ parameterlength(a::AbstractArray) = length(a)
 Return the total number of states of the layer `l`.
 """
 statelength(l::AbstractExplicitLayer) = statelength(initialstates(Random.default_rng(), l))
+statelength(l::AbstractExplicitLayer{false, hasparams}) where {hasparams} = 0
```
Comment (avik-pal) [inline src/core.jl:78]: Ordering

File: src/core.jl [hunk 5]
```diff
@@ -83,6 +100,30 @@ function apply(model::AbstractExplicitLayer, x, ps, st::NamedTuple)
...
+function apply(model::AbstractExplicitLayer{true, false}, x, ps)
+    return apply(model, x, ps, NamedTuple())
+end
+
+function apply(model::AbstractExplicitLayer{false, true}, x, st::NamedTuple)
+    return apply(model, x, NamedTuple(), st)
+end
+
+function apply(model::AbstractExplicitLayer{false, false}, x)
+    return apply(model, x, NamedTuple(), NamedTuple())
```
Comment (avik-pal) [inline src/core.jl:112]: I think there is a disconnect between these functions and the original goal. The idea of having these properties in the type, would be to simplify user code. So a stateless layer can define `(::Layer)(x, ps)`. With this the user is still forced to define the original 3 argument api.

File: src/core.jl [hunk 6]
```diff
@@ -91,36 +132,68 @@ end
...
+              AbstractExplicitLayer{hasparams, hasstate} end
 
 function initialparameters(rng::AbstractRNG,
                            l::AbstractExplicitContainerLayer{layers}) where {layers}
     length(layers) == 1 && return initialparameters(rng, getfield(l, layers[1]))
     return NamedTuple{layers}(initialparameters.(rng, getfield.((l,), layers)))
 end
 
+function initialparameters(::AbstractRNG,
+                           ::AbstractExplicitContainerLayer{layers, false, hasstate}) where {
```
Comment (avik-pal) [inline src/core.jl:154]: unused `hasstate`

File: src/core.jl [hunk 7]
```diff
@@ -91,36 +132,68 @@ end
...
+                                                                                          hasparams
+                                                                                          }
+    return NamedTuple()
+end
+
 function parameterlength(l::AbstractExplicitContainerLayer{layers}) where {layers}
     return sum(parameterlength, getfield.((l,), layers))
 end
 
+function parameterlength(::AbstractExplicitContainerLayer{layers, false, hasstate}) where {
```
Comment (avik-pal) [inline src/core.jl:179]: unused `hasstate`

**Review by avik-pal — CHANGES_REQUESTED** [review]
Made a first pass over the PR.

**Comment by avik-pal** [comment]
Also if we are to simplify user code, we will have to define functions like `(l::AbstractExplicitLayer{false}(x, ps, st) = l(x, st)`,  `(l::AbstractExplicitLayer{hasparams, false}(x, ps, st) = l(x, ps), st`, `(l::AbstractExplicitLayer{false, false}(x, ps, st) = l(x), st`

**Comment by avik-pal** [comment]
Also tagging @chrisrackauckas, @touchesir for thoughts on this

**Comment by avik-pal** [comment]
>  One would need to apply a logical OR to the hasparams / hasstate of each element of the NamedTuple, which sounds like it has to be done dynamically since we don't even know its length.

That information is available during compilation, I don't think it will lead to a dynamic dispatch. Worst case if the compiler can't do that, we can just use a generated function. Also, I wouldn't care too much about it since it is never going to appear in a performance-critical section of the code.

**Comment by ToucheSir** [comment]
I don't see any scenario where you'd need a generated function either, as the aforementioned bool type logic happens at construction. Have you considered whether it makes sense to add type aliases for some of the more common combinations?

**Comment by ChrisRackauckas** [comment]
I'm not sure I like the high level call changes with this. There's something simple about always needing `f(x,p,st)`. If you start being able to drop some in some cases, then a lot of docs will be like "oh yeah it's just `f(x,p)`" (most SciML examples), but then the user goes to put in a BatchNorm and suddenly their code throws an odd error which they don't understand. I think there's simplicity in having only one calling convention.

File: src/core.jl (hunk 1 above)
Comment (gdalle) [reply src/core.jl]: Fixed in [37f4851](https://github.com/avik-pal/Lux.jl/pull/84/commits/37f48519a731897b9de5771ccb714ff0eea2e013)

File: src/core.jl (hunk 3 above)
Comment (gdalle) [reply src/core.jl]: Done in [37f4851](https://github.com/avik-pal/Lux.jl/pull/84/commits/37f48519a731897b9de5771ccb714ff0eea2e013)

File: src/core.jl (hunk 4 above)
Comment (gdalle) [reply src/core.jl]: Fixed in [37f4851](https://github.com/avik-pal/Lux.jl/pull/84/commits/37f48519a731897b9de5771ccb714ff0eea2e013)

File: src/core.jl (hunk 2 above)
Comment (gdalle) [reply src/core.jl]: Fixed in [37f4851](https://github.com/avik-pal/Lux.jl/pull/84/commits/37f48519a731897b9de5771ccb714ff0eea2e013)

File: src/core.jl (hunk 6 above)
Comment (gdalle) [reply src/core.jl]: Fixed in [37f4851](https://github.com/avik-pal/Lux.jl/pull/84/commits/37f48519a731897b9de5771ccb714ff0eea2e013)

File: src/core.jl (hunk 7 above)
Comment (gdalle) [reply src/core.jl]: Fixed in [37f4851](https://github.com/avik-pal/Lux.jl/pull/84/commits/37f48519a731897b9de5771ccb714ff0eea2e013)

**Comment by gdalle** [comment]
> I'm not sure I like the high level call changes with this. 
@ChrisRackauckas I think that depending on your use of the library, those changes might make a lot of sense. For instance, the notion of layer state is not obvious to beginners, and when they don't need it (eg. for a simple CNN without dropout), carrying it along makes the code heavier and less readable.
As mentioned by @avik-pal in #83, it's mostly about quality of life, and to help people get started with Lux.jl. Of course, as soon as you get more familiar with the syntax, you will start adding a state everywhere. But to ease first contact with the code, leaving the state aside for simple networks is a clear bonus to me.

**Comment by gdalle** [comment]
> That information is available during compilation, I don't think it will lead to a dynamic dispatch. Worst case if the compiler can't do that, we can just use a generated function. Also, I wouldn't care too much about it since it is never going to appear in a performance-critical section of the code.

I tried defining the parameters as follows:
```julia
P = maximum(hasparams, layers)
S = maximum(hasstate, layers)
```
but for some reason it messed up type inference for the `NamedTuple` layer field, which made all the tests fail due to the generated functions that are used to apply eg. `Parallel`.

I'll come back to that later.

**Comment by gdalle** [comment]
> Also, I wouldn't care too much about it since it is never going to appear in a performance-critical section of the code.

I'm not sure I agree. There are cases where you want to create a neural network inside a hot loop, and faulty type inference at construction can mess up everything that follows.

**Comment by ChrisRackauckas** [comment]
As I mentioned, no example I have written (which now is like 30 tutorials or so that use Lux) would require state. So the SciML uses of the library would definitely fall into the category that you are trying to target. But even with that, I don't think it's a good idea because this is so pervasive that I think it will make the cases with state confusing. If no tutorials ever include the state, then how are you ever expected to know that it exists? What I foresee happening is that there will be some layers that users will get confused by, like BatchNorm will just throw "mysterious" errors, when in reality the mystery is that the tutorials are all focused on a special case. 

Now this is somewhat hidden in some cases like NeuralPDE and DiffEqFlux  since in those cases users are rarely doing the call themselves. But SciMLSensitivity will definitely have this as an extra documentation issue, and either the solution will be to change our tutorials to showcase state in what will then be a non-standard way (i.e. include state even when it's nothing), always use layers with state in tutorials as a matter of training for the most "robust" form, or just ignoring it as a Lux issue and answering a bunch of issues with "oh, that's a Lux thing. If you use xxx then you have `st` and ...". I hate doing the latter because it quickly leads to 20 emails to answer everyday with the same thing, and by now I've learned that styles which require answering people's questions are just not scalable.

**Comment by gdalle** [comment]
> What I foresee happening is that there will be some layers that users will get confused by, like BatchNorm will just throw "mysterious" errors

I also think that this is a foreseeable problem, but one that can be fixed with explicit error messages on the Lux.jl side when the wrong methods are called. Something like
```
Your neural network contains a layer that relies on a state variable. You probably tried to call it with `network(x, ps)`, but instead you need to use `network(x, ps, st)`. See [...] in the documentation for more details.
```

Anyway, I only started working on this because @avik-pal agreed it would be a welcome feature. If he ends up agreeing it's not, then I'll simply close the PR

**Comment by gdalle** [comment]
Another justification for this feature would be easing the transition from Flux.jl to Lux.jl. Since you announced that SciML would be using Lux.jl, I've been convinced that this transition is a good thing. And to get people hooked to Lux.jl, the fewer visible differences with Flux.jl, the better. Then, when models get more complex and they're in too deep, we can drop the bomb with the additional state argument and they won't be able to escape :sweat_smile:

**Comment by ChrisRackauckas** [comment]
The transition is pretty much done BTW, so check out

- https://sensitivity.sciml.ai/dev/
- https://diffeqflux.sciml.ai/dev/
- https://neuralpde.sciml.ai/dev/

which all bumped majors and removed `FastChain` from everything (well, DiffEqFlux is keeping the depwarns around for awhile to help the updates). 🎉 

Almost everything is Lux now (https://github.com/SciML/NeuralPDE.jl/issues/533 and https://github.com/SciML/NeuralPDE.jl/issues/533 pretty much sealed the deal), except a notable issue with Lux on GPUs (due to ComponentArrays not broadcasting on GPUs correctly). There is an on-going thread about if we're happy with the new coding styles these updates gave, so I encourage you to chime in (https://julialang.slack.com/archives/CN04R7WKE/p1657113244013009). 

Indeed, one of the things that was missing from `FastChain` was state, so when doing these updates we had to add state in. And right now it doesn't look nice, it's using `global`s, which Chris Elrod wasn't too keen on, and there are some `@ignore_derivatives` points which seem to be correct for every case I discussed with @avik-pal though I do wonder whether derivatives w.r.t. state are something that may be required for something (and would love to hear if there is a case). For some things like `NeuralODE` it would be good to disallow models that have some form of state that we know would be incorrect (for example, LSTMs in a `NeuralODE` are bad because the neural ODE doesn't step forward, and so the adaptivity means an LSTM-defined neural ODE does not have a well-defined solution if it rejects steps!), but disallowing all models with state goes too far (normalization layers for example should be fine in the theory that the state is fixed for each ODE solve).

However, even with all of those new issues arising because of state, I think we need to just fine what coding behaviors and habits do it well. For example, always make a callable struct to enclose the state in an ODE definition? We're still working it out, but I think it would be best to find what the nice styles are here and use it everywhere. Otherwise it might just be one "advanced" example that shows it, with other examples showing "worse" coding styles, or even worse, coding styles that have some unnecessary complexity when no state is there (but done anyways for consistency with the cases where it is?).

That said, I will be willing to give it a try since it would simplify docs and such, though I am worried that in all tutorials of these 3 libraries, we would just remove state to match this PR. And if we do that, 😅 state examples will seem like magic. A good error message may be enough though.

**Comment by gdalle** [comment]
> The transition is pretty much done BTW

I know the transition is done for SciML (kudos on the massive change!), but there's lots of other people using Flux out there that we could still convince to switch :wink: 

> That said, I will be willing to give it a try since it would simplify docs and such, though I am worried that in all tutorials of these 3 libraries, we would just remove state to match this PR.

I'm not necessarily advocating for the `layer(x, ps)` to become the default (recommended) syntax, as a matter of fact I'm not sure it should.
And as I stated in [this comment](https://github.com/avik-pal/Lux.jl/pull/84#discussion_r917362838), I think that for people who _define_ new layers, sticking to the three-argument syntax makes more sense.
The change I'm proposing is mostly for people who _use_ pre-defined stateless layers, to which we can give additional options.

**Comment by avik-pal** [comment]
I kind of agree with Chris here that having a consistent API, even if it is not beginner-friendly, might be the way to go. Flux is great in that you can pretty much throw anything its way, and it will work, but for Lux, that is not the case. It relies on a pretty strict convention on how you should be writing your julia/zygote code. 

Originally I was thinking about how to make it easier for people writing custom layers, in which case managing states is quite cumbersome and, in most cases, can be auto-generated. See #46, which kind of already does that (my only concern with merging that PR even though its nearly complete, users will not check the generated code and slap it around places where it might not work correctly)

Also in my experience, it doesn't matter if you show users the "correct" way to do something (especially after introducing a partially correct easier version). They tend to gravitate towards the easier (and in this particular case) the wrong way to do things. Anecdotally, flax which also operates similarly to lux (actually the other way around), introduces a non-state managed API first. And even after two months of using it, I still can't remember how to handle states properly.

**Comment by gdalle** [comment]
Okay, so do we just close this and move on?

**Comment by gdalle** [comment]
I must say the discussion also swayed me towards keeping the state around

**Comment by avik-pal** [comment]
> Okay, so do we just close this and move on?

I think yeah. Also, I should have tagged people in the other issue before you put all this effort in. My bad.
